### PR TITLE
fix: clickBookmark's DB query was not migrated

### LINKED ISSUE #389: /nextcloud/app/bookmarks/bookmark/click HTTP 500
### Steps to reproduce
1. Open Bookmarks app

### Expected behaviour
App loads without errors in access logs.

### Actual behaviour
HTTP 500 error for /nextcloud/apps/bookmarks/bookmark/click (see server/nextcloud logs below).

### Solution
I was able to fix this by installing `libjs-backbone`, but wanted to report it to see if I installed it incorrectly (from the app store) or if perhaps that needed to be added to the documentation (or OC/NC docs?) as a dependency.

### Server configuration
**Operating system**: Ubuntu 16.04

**Web server:** Apache (apache2handler)

**Database:** mysql 5.7.18

**PHP version:** 7.0.18-0ubuntu0.16.04.1

Modules loaded: Core, date, libxml, openssl, pcre, zlib, filter, hash, Reflection, SPL, session, standard, apache2handler, mysqlnd, PDO, xml, bcmath, bz2, calendar, ctype, curl, dom, mbstring, fileinfo, ftp, gd, gettext, iconv, igbinary, imagick, imap, intl, json, ldap, exif, mcrypt, memcached, mongodb, mysqli, pdo_mysql, pdo_pgsql, pdo_sqlite, pgsql, Phar, posix, readline, redis, shmop, SimpleXML, smbclient, sockets, sqlite3, sysvmsg, sysvsem, sysvshm, tidy, tokenizer, wddx, xmlreader, xmlrpc, xmlwriter, xsl, zip, libsmbclient, Zend OPcache

**Nextcloud version:** 12.0.0 - 12.0.0.29

**Updated from an older Nextcloud/ownCloud or fresh install:**
Updated

**Where did you install Nextcloud from:** 
zip

**Signing status:**
```
[]
```

**List of activated apps:**

```
Enabled:
 - activity: 2.5.2
 - admin_audit: 1.2.0
 - bookmarks: 0.10.0
 - bruteforcesettings: 1.0.2
 - calendar: 1.5.3
 - comments: 1.2.0
 - contacts: 1.5.3
 - dav: 1.3.0
 - federatedfilesharing: 1.2.0
 - federation: 1.2.0
 - files: 1.7.2
 - files_antivirus: 1.0.0.0
 - files_automatedtagging: 1.2.2
 - files_opds: 0.8.2
 - files_pdfviewer: 1.1.1
 - files_reader: 1.0.4
 - files_sharing: 1.4.0
 - files_snapshots: 0.1.1
 - files_texteditor: 2.4.1
 - files_trashbin: 1.2.0
 - files_versions: 1.5.0
 - files_videoplayer: 1.1.0
 - firstrunwizard: 2.1
 - gallery: 17.0.0
 - impersonate: 1.0.1
 - issuetemplate: 0.2.1
 - keeweb: 0.4.0
 - logreader: 2.0.0
 - lookup_server_connector: 1.0.0
 - nextcloud_announcements: 1.1
 - notifications: 2.0.0
 - oauth2: 1.0.5
 - ownnote: 1.08
 - password_policy: 1.2.2
 - provisioning_api: 1.2.0
 - serverinfo: 1.2.0
 - sharebymail: 1.2.0
 - survey_client: 1.0.0
 - systemtags: 1.2.0
 - tasks: 0.9.5
 - theming: 1.3.0
 - twofactor_backupcodes: 1.1.1
 - updatenotification: 1.2.0
 - user_external: 0.4
 - workflowengine: 1.2.0

Disabled:
 - encryption
 - files_antivirus.old
 - files_external
 - richdocuments
 - spreedme
 - templateeditor
 - user_ldap
```

**The content of config/config.php:**

```
{
    "instanceid": "oc2p2s8mi80a",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "***REMOVED SENSITIVE VALUE***",
        "***REMOVED SENSITIVE VALUE***"
    ],
    "updater.server.url": "https:\/\/updates.nextcloud.org\/updater_server\/",
    "datadirectory": "\/media\/ext\/.nextcloud\/data",
    "overwrite.cli.url": "https:\/\/example.com\/nextcloud",
    "htaccess.RewriteBase": "\/nextcloud",
    "dbtype": "mysql",
    "version": "12.0.0.29",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "localhost",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "logtimezone": "UTC",
    "installed": true,
    "theme": "",
    "maintenance": false,
    "loglevel": 0,
    "appstore.experimental.enabled": true,
    "updater.release.channel": "stable",
    "updater.secret": "***REMOVED SENSITIVE VALUE***",
    "defaultapp": "tasks",
    "trusted_proxies": [
        "127.0.0.1"
    ],
    "forwarded_for_headers": [
        "HTTP_X_FORWARDED_FOR"
    ],
    "auth.bruteforce.protection.enabled": true,
    "enabledPreviewProviders": [
        "OC\\Preview\\Image",
        "OC\\Preview\\MP3",
        "OC\\Preview\\TXT",
        "OC\\Preview\\MarkDown",
        "OC\\Preview\\Epub",
        "OC\\Preview\\PDF",
        "OC\\Preview\\OpenDocument",
        "OC\\Preview\\StarOffice",
        "OC\\Preview\\MSOfficeDoc",
        "OC\\Preview\\MSOffice2003",
        "OC\\Preview\\MSOffice2007",
        "OC\\Preview\\FB2"
    ],
    "mail_smtpmode": "php",
    "memcache.local": "OC\\Memcache\\Redis",
    "memcache.locking": "OC\\Memcache\\Redis",
    "redis": {
        "host": "localhost",
        "port": 6379,
        "dbindex": 0,
        "timeout": 1.5
    },
    "mail_smtpsecure": "tls",
    "mail_smtpauthtype": "LOGIN",
    "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpport": "587",
    "mail_smtpauth": 1,
    "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
    "mail_smtppassword": "***REMOVED SENSITIVE VALUE***"
}
```

**Are you using external storage, if yes which one:** files_external is disabled

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** No

### Client configuration
**Browser:** Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/58.0.3029.110 Chrome/58.0.3029.110 Safari/537.36

**Operating system:** Ubuntu 16.04

#### Web server error log

```
xxx.xxx.xxx.xxx - - [30/Jun/2017:13:41:06 -0500] "POST /nextcloud/apps/bookmarks/bookmark/click HTTP/1.1" 500 7419 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/58.0.3029.110 Chrome/58.0.3029.110 Safari/537.36"
```

#### Nextcloud log (data/nextcloud.log)
```
{"reqId":"WVabQn8AAQEAABh1xMoAAAAB","level":3,"time":"2017-06-30T18:41:06+00:00","remoteAddr":"192.168.99.105","user":"exampleuser","app":"index","method":"POST","url":"\/nextcloud\/apps\/bookmarks\/bookmark\/click","message":"Exception: {\"Exception\":\"Error\",\"Message\":\"Call to undefined method OC\\\\DB\\\\Connection::prepareQuery()\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/bookmarks\\\/controller\\\/rest\\\/internalbookmarkcontroller.php(132): OCA\\\\Bookmarks\\\\Controller\\\\Rest\\\\BookmarkController->clickBookmark('http:\\\/\\\/arpon.so...')\\n#1 [internal function]: OCA\\\\Bookmarks\\\\Controller\\\\Rest\\\\InternalBookmarkController->clickBookmark('http:\\\/\\\/arpon.so...')\\n#2 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(160): call_user_func_array(Array, Array)\\n#3 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(90): OC\\\\AppFramework\\\\Http\\\\Dispatcher->executeController(Object(OCA\\\\Bookmarks\\\\Controller\\\\Rest\\\\InternalBookmarkController), 'clickBookmark')\\n#4 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/AppFramework\\\/App.php(114): OC\\\\AppFramework\\\\Http\\\\Dispatcher->dispatch(Object(OCA\\\\Bookmarks\\\\Controller\\\\Rest\\\\InternalBookmarkController), 'clickBookmark')\\n#5 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/AppFramework\\\/Routing\\\/RouteActionHandler.php(47): OC\\\\AppFramework\\\\App::main('InternalBookmar...', 'clickBookmark', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#6 [internal function]: OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler->__invoke(Array)\\n#7 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Route\\\/Router.php(299): call_user_func(Object(OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler), Array)\\n#8 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/base.php(1000): OC\\\\Route\\\\Router->match('\\\/apps\\\/bookmarks...')\\n#9 \\\/var\\\/www\\\/html\\\/nextcloud\\\/index.php(40): OC::handleRequest()\\n#10 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/bookmarks\\\/controller\\\/rest\\\/bookmarkcontroller.php\",\"Line\":297}","userAgent":"Mozilla\/5.0 (X11; Linux x86_64) AppleWebKit\/537.36 (KHTML, like Gecko) Ubuntu Chromium\/58.0.3029.110 Chrome\/58.0.3029.110 Safari\/537.36","version":"12.0.0.29"}
```

#### Browser log

```
N/A
```

### PR DESCRIPTION
**Pull request by blizzz**
Fixes #389

### CODE REVIEW

File: controller/rest/bookmarkcontroller.php [hunk 1]
```diff
@@ -307,20 +307,17 @@ public function deleteBookmark($id = -1) {
...
-				AND `url` LIKE ?
-		');
-
-		$params = array($this->userId, htmlspecialchars_decode($url));
-		$query->execute($params);
+		$qb = $this->db->getQueryBuilder();
+		$qb->update('bookmarks')
+			->set('clickcount', $qb->createFunction('`clickcount` +1'))
+			->where($qb->expr()->eq('user_id', $qb->createNamedParameter($this->userId)))
+			->andWhere($qb->expr()->like('url', $qb->createNamedParameter(htmlspecialchars_decode($url))))
```
Comment (marcelklehr) [inline controller/rest/bookmarkcontroller.php]: Should we use escapeLikeParameter here, too? Might be a good idea, since urls tend to contain`%`s.

Comment (blizzz) [inline controller/rest/bookmarkcontroller.php]: Good thought and idea! When trying with an address containing an & I also figured out that decoding is necessary.  Fixed, with tests.

**Comment by marcelklehr** [comment]
Nice!

Mh. Weird. The new test fails unpredictably :/

**Comment by blizzz** [comment]
Not so unpredictable, stable9 and stable10 are faling, also stable11 on sqlite. Probably due to missing ``ESCAPE '\'``. Let's see whether this attempt fixes it :)

**Comment by blizzz** [comment]
does not seem so :( try to figure it out tonight

**Comment by blizzz** [comment]
it's flying ✈️ 🕺

**Comment by marcelklehr** [comment]
:birthday:

**Pull request merged**